### PR TITLE
[TensorRT EP] Update TRT10.0 deprecated api 

### PR DESCRIPTION
**Pull request by yf711**
### Description

Note:
* This PR would remove C4996 suppression in tensorrt_execution_provider.cc only (according to Nvidia, places with nvinfer.h included need C4996 suppression, when /Zc:__cplusplus is enabled in ORT win build)
* A follow-up PR will be raised to update deprecated TRT Plugin api usage.

Here are deprecated apis to be updated in this PR:
| deprecated api                                               | Update                                                       |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| [kCUBLAS](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/namespacenvinfer1.html#a9e1d81e5a8bfeb38b86e22a66d5f836a) | /                                                            |
| [kCUBLAS_LT](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/namespacenvinfer1.html#a9e1d81e5a8bfeb38b86e22a66d5f836a) | /                                                            |
| [kCUDNN](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/namespacenvinfer1.html#a9e1d81e5a8bfeb38b86e22a66d5f836a) | /                                                            |
| [reallocateOutput](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1v__1__0_1_1_i_output_allocator.html#acae6441d4029584cc1c6550917518691) | Superseded by [reallocateOutputAsync](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1v__1__0_1_1_i_output_allocator.html#aa40eeb891c1dfe4c1bbf1eabe8c705ab) with cudaStream_t argument |
| [createExecutionContextWithoutDeviceMemory](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_cuda_engine.html#adc86bcc42b098204997396ef2b1093fb) | Superseded by [createExecutionContext()](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_cuda_engine.html#a35de29aa6134165a5b14a537e6d99e82) with parameter.<br />Check [ExecutionContextAllocationStrategy::kUSER_MANAGED](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/namespacenvinfer1.html#ac6251a050df629edfc0ce037fa366503) for more detail |

### Motivation and Context

TRT deprecated api list: https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/deprecated.html

### CODE REVIEW

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 1]
```diff
@@ -289,10 +298,30 @@ void CudaCall<cudnnStatus_t, true>(cudnnStatus_t retCode, const char* exprString
   return g_host->CudaCall_true(retCode, exprString, libName, successCode, msg, file, line);
 }
 
+#if NV_TENSORRT_MAJOR > 8
```
Comment (chilo-ms) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:301]: Make this MACRO consistent with the MACRO in the tensorrt_execution_provider.h where you use NV_TENSORRT_MAJOR >= 10

Comment (yf711) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: That is fixed. Thanks!

**Comment by tianleiwu** [comment]
Please remove lines with `4996` from the following files. 
* onnxruntime/core/providers/tensorrt/nv_includes.h
* onnxruntime/test/unittest_main/test_main.cc

**Comment by yf711** [comment]
> Please remove lines with `4996` from the following files.
> 
> * onnxruntime/core/providers/tensorrt/nv_includes.h
> * onnxruntime/test/unittest_main/test_main.cc

Thanks for the context. These 4996 usage have been reverted

**Review by chilo-ms — APPROVED** [review]

**Comment by jywu-msft** [comment]
there are build errors?
 
D:\a\_work\_temp\TensorRT-10.0.1.6.Windows10.x86_64.cuda-11.8\include\NvInferRuntimeBase.h#L593
D:\a\_work\_temp\TensorRT-10.0.1.6.Windows10.x86_64.cuda-11.8\include\NvInferRuntimeBase.h(593,16): Error C4996: 'nvinfer1::v_1_0::IGpuAllocator::allocate': was declared deprecated

**Comment by yf711** [comment]
Including nvinfer.h would cause C4996 errors during Windows ORT build, when `/Zc:__cplusplus` is enabled:
nvinfer1::v_1_0::IGpuAllocator::allocate
nvinfer1::v_1_0::IGpuAllocator::deallocate
nvinfer1::IPluginV2
nvinfer1::IPluginV2Ext 
It seems that there are indirect usage of these api by TRT internally. I'm adding C4996 suppression back temporarily and checking with Nvidia for better solution on that.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_custom_ops.cc [hunk 2]
```diff
@@ -60,18 +60,23 @@ common::Status CreateTensorRTCustomOpDomainList(std::vector<OrtCustomOpDomain*>&
...
+    auto plugin_creators = getPluginRegistry()->getAllCreators(&num_plugin_creator);
+#else
     auto plugin_creators = getPluginRegistry()->getPluginCreatorList(&num_plugin_creator);
+#endif
     std::unordered_set<std::string> registered_plugin_names;
 
     for (int i = 0; i < num_plugin_creator; i++) {
+#if NV_TENSORRT_MAJOR >= 10
+      auto plugin_creator_interface = plugin_creators[i];
+      auto plugin_creator = static_cast<nvinfer1::IPluginCreator*>(plugin_creator_interface);
```
Comment (chilo-ms) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_custom_ops.cc:74]: nvinfer1::IPluginCreator is deprecated in TRT 10 as well

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 3]
```diff
@@ -289,10 +298,30 @@ void CudaCall<cudnnStatus_t, true>(cudnnStatus_t retCode, const char* exprString
...
+  // if cudaMalloc fails, returns nullptr.
+  return outputPtr;
+}
+#else
+// Only override this method when TensorRT <= 8.6
 void* OutputAllocator::reallocateOutput(char const* /*tensorName*/, void* /*currentMemory*/, uint64_t size,
                                         uint64_t /*alignment*/) noexcept {
   // Some memory allocators return nullptr when allocating zero bytes, but TensorRT requires a non-null ptr
   // even for empty tensors, so allocate a dummy byte.
+  LOGS_DEFAULT(WARNING) << "[TensorRT EP] reallocateOutput() is deprecated after TensorRT 8.6";
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:324]: remove this warning log. it's not useful or actionable by the user.

**Comment by jywu-msft** [comment]
was the rellocateOutput change tested on Faster-RCNN with oss parser?

**Comment by yf711** [comment]
> was the rellocateOutput->reallocateOutputAsync change tested on Faster-RCNN with oss parser?

That has been tested with no issue

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc (hunk 3 above)
Comment (yf711) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: This warning log has been removed

**Pull request merged**